src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
+
+    for (TopicPartition topicPart : writers.keySet()) {
       rotate(topicPart);
     }
     writers.clear();
+
+    for (WAL wal : wals.values()) {
+      wal.close();
+    }
+    wals.clear();
+
+    tempFileNames.clear();
+    recordCounters.clear();
+    offsets.clear();
+    tasks.clear();
     if (fs != null) {
       fs.close();
     }
   }
 
-  private Map<TopicPartition, Long> readOffsets() throws IOException {
-    Map<TopicPartition, Long> offsets = new HashMap<TopicPartition, Long>();
+  private void recover(TopicPartition topicPart) throws IOException {
+    if (!recovered.contains(topicPart)) {
+      WAL wal = new WAL(conf, url, topicsDir, topicPart);
+      wal.apply();
+      wals.put(topicPart, wal);
+      recovered.add(topicPart);
+      readOffsets(topicPart);
+    }
+  }
+
+  private void readOffsets() throws IOException {
     Path path = new Path(url + "/" + topicsDir);
-    PathFilter filter = new CommittedFileFilter();
     FileStatus[] topics = fs.listStatus(path);
-    for (FileStatus topic: topics) {
+    for (FileStatus topic : topics) {
       if (topic.isDirectory()) {
         String topicName = topic.getPath().getName();
-        FileStatus[] partitions = fs.listStatus(topic.getPath(), filter);
+        FileStatus[] partitions = fs.listStatus(topic.getPath());
         for (FileStatus partition : partitions) {
-          if (partition.isFile()) {
-            String filename = partition.getPath().getName();
-            String[] parts = filename.split("\\.");
-            try {
-              int partitionNum = Integer.parseInt(parts[0]);
-              long offset = Long.parseLong(parts[1]);
-              TopicPartition topicPart = new TopicPartition(topicName, partitionNum);
-              if (!offsets.containsKey(topicPart) || offset > offsets.get(topicPart)) {
-                offsets.put(topicPart, offset);
-              }
-            } catch (NumberFormatException e) {
-              log.warn("Invalid Copycat HDFS file: {}", filename);
-            }
+          if (partition.isDirectory()) {
+            int partitionNum = Integer.parseInt(partition.getPath().getName());
+            TopicPartition topicPart = new TopicPartition(topicName, partitionNum);
+            readOffsets(topicPart);
           }
         }
       }
     }
-    return offsets;
   }
 
+  private void readOffsets(TopicPartition topicPart) throws IOException {
+    String topic = topicPart.topic();
+    int partition = topicPart.partition();
+    Path path = new Path(url + "/" + topicsDir + "/" + topic + "/" + partition);
+    if (!fs.exists(path)) {
+      return;
+    }
+    PathFilter filter = new CommittedFileFilter();
+    FileStatus[] committedFiles = fs.listStatus(path, filter);
+    for (FileStatus committedFile : committedFiles) {
+      String filename = committedFile.getPath().getName();
+      String[] parts = filename.split("_");
+      try {
+        long endOffset = Long.parseLong(parts[1]);
```

Comment:
We don't need to catch IndexOutOfBoundsException here as the CommittedFileFilter ensures that only valid committed files are returned for a listStatus call.  I agree that we need to add tests to make sure that CommittedFileFilter behaves properly.